File: go/internal/ingest/ubx.go
package ingest

import (
	"bufio"
	"encoding/binary"
	"io"
	"math"
	"time"

	"github.com/ptudor/gnss"
	"github.com/ptudor/navlistener/internal/metrics"
)

// UBX protocol constants (u-blox interface description). SFRBX carries raw nav
// subframe words per gnssId/sigId (docs/CONSTELLATIONS.md §2.1).
const (
	ubxSync1      = 0xB5
	ubxSync2      = 0x62
	ubxClassNAV   = 0x01
	ubxIDNAVSAT   = 0x35
	ubxClassRXM   = 0x02
	ubxIDSFRBX    = 0x13
	ubxIDRAWX     = 0x15
	ubxClassMON   = 0x0A
	ubxIDMONHW    = 0x09
	ubxIDMONRF    = 0x38
	ubxMaxPayload = 1 << 14 // guard against a corrupt length prefix
)

// scanUBX reads a UBX byte stream and calls emit for each UBX-RXM-SFRBX message,
// converted to a RawFrame. It resynchronises on the 0xB5 0x62 sync pattern and
// drops any message whose Fletcher checksum fails, so a mid-stream connect or a
// corrupt frame never derails the reader. It returns when the reader errors (EOF /
// connection drop), which the connector treats as a reconnect trigger.
func scanUBX(r io.Reader, source string, now func() time.Time, emit func(*RawFrame), onErr func(kind string)) error {
	br := bufio.NewReaderSize(r, 1<<16)
	for {
		// Resynchronise to the sync pattern.
		if err := syncTo(br, ubxSync1, ubxSync2); err != nil {
			return err
		}
		// peek the header+payload+checksum without consuming it. On a
		// length or checksum failure, nothing here is Discarded, so the next
		// syncTo call resumes scanning byte-by-byte from right after the sync
		// pattern instead of skipping the whole claimed (possibly bogus) extent
		// (up to ubxMaxPayload+2 = 16386 bytes), which may contain a real frame.
		hdr, err := br.Peek(4) // class, id, length(2, LE)
		if err != nil {
			return err
		}
		length := int(binary.LittleEndian.Uint16(hdr[2:]))
		if length > ubxMaxPayload {
			onErr("ubx_length") // implausible length: skip and resync
			continue
		}
		total := 4 + length + 2 // header + payload + 2 checksum bytes
		peeked, err := br.Peek(total)
		if err != nil {
			return err
		}
		body := peeked[4 : 4+length]
		ckA, ckB := fletcher8(peeked[0], peeked[1], peeked[2], peeked[3], body)
		if ckA != peeked[total-2] || ckB != peeked[total-1] {
			onErr("ubx_checksum")
			continue
		}
		// Copy what's needed before Discard invalidates br's buffer view (peeked
		// aliases it and is only valid until the next read/Discard).
		cls, id := peeked[0], peeked[1]
		body = append([]byte(nil), body...)
		if _, err := br.Discard(total); err != nil {
			return err
		}
		switch {
		case cls == ubxClassRXM && id == ubxIDSFRBX:
			if f := parseSFRBX(body, source, now()); f != nil {
				emit(f)
			} else {
				onErr("ubx_sfrbx")
			}
		case cls == ubxClassRXM && id == ubxIDRAWX:
			// only STRUCTURAL malformation is a parse error. A well-formed
			// RAWX with nothing usable — cold-start week==0/invalid rcvTow, or every
			// measurement's PR-valid bit clear — is a normal receiver start-up state,
			// already accounted per-field by RawObsInvalidTotal; counting it here too
			// inflated IngestErrorsTotal{ubx_rawx} through every cold/warm start (and
			// double-counted the week==0 case).
			if _, ok := parseRAWX(body, source, now(), emit); !ok {
				onErr("ubx_rawx")
			}
		case cls == ubxClassMON && id == ubxIDMONRF:
			if f := parseMONRF(body, source, now()); f != nil {
				emit(f)
			} else {
				onErr("ubx_monrf")
			}
		case cls == ubxClassMON && id == ubxIDMONHW:
			if f := parseMONHW(body, source, now()); f != nil {
				emit(f)
			} else {
				onErr("ubx_monhw")
			}
		case cls == ubxClassNAV && id == ubxIDNAVSAT:
			if f := parseNAVSAT(body, source, now()); f != nil {
				emit(f)
			} else {
				onErr("ubx_navsat")
			}
		}
	}
}

// parseRAWX converts a UBX-RXM-RAWX payload into one observation RawFrame per
// measurement (u-blox interface description: 16-byte header — rcvTow R8 in
// seconds of week, week U2, leapS I1, numMeas U1, recStat X1 — then 32 bytes per
// measurement). These are the dual-frequency observables feeding the measured
// ionosphere (docs/MATH.md §7.4); they are telemetry, not nav frames, so they
// carry Obs instead of Words. Returns the number of measurements emitted and
// whether the payload was structurally well-formed : ok=false only for
// bytes that cannot carry a valid RAWX (short header, numMeas overrunning the
// payload) — a checksum-valid message shaped like that means the receiver/link
// really did corrupt something. A structurally-valid payload with no usable
// measurements returns (0, true).
func parseRAWX(p []byte, source string, recv time.Time, emit func(*RawFrame)) (int, bool) {
	if len(p) < 16 {
		return 0, false
	}
	rcvTow := math.Float64frombits(binary.LittleEndian.Uint64(p[0:]))
	week := int(binary.LittleEndian.Uint16(p[8:]))
	numMeas := int(p[11])
	if 16+numMeas*32 > len(p) {
		return 0, false
	}
	if !finiteFloat(rcvTow) || rcvTow < 0 || rcvTow >= 604800 || week == 0 {
		metrics.RawObsInvalidTotal.WithLabelValues(source, "time").Inc()
		return 0, true
	}
	emitted := 0
	for i := 0; i < numMeas; i++ {
		m := p[16+i*32:]
		prM := math.Float64frombits(binary.LittleEndian.Uint64(m[0:]))
		cpCyc := math.Float64frombits(binary.LittleEndian.Uint64(m[8:]))
		doHz := math.Float32frombits(binary.LittleEndian.Uint32(m[16:]))
		trkStat := m[30]
		// trkStat bit 0 = pseudorange valid, bit 1 = carrier phase valid.
		if trkStat&0x01 == 0 {
			continue
		}
		if !finiteFloat(prM) || prM <= 0 || prM > 1e9 {
			metrics.RawObsInvalidTotal.WithLabelValues(source, "pseudorange").Inc()
			continue
		}
		if !finiteFloat(float64(doHz)) || math.Abs(float64(doHz)) > 1e6 {
			metrics.RawObsInvalidTotal.WithLabelValues(source, "doppler").Inc()
			continue
		}
		cpValid := trkStat&0x02 != 0
		arcBreak := false
		if cpValid && (!finiteFloat(cpCyc) || math.Abs(cpCyc) > 1e10) {
			metrics.RawObsInvalidTotal.WithLabelValues(source, "carrier").Inc()
			cpValid, arcBreak = false, true
			cpCyc = 0
		}
		emit(&RawFrame{
			Recv:   recv,
			Source: source,
			GnssID: gnss.GNSSID(m[20]),
			SvID:   int(m[21]),
			SigID:  int(m[22]),
			FreqID: int(m[23]),
			Obs: &RawObs{
				RcvTow:     rcvTow,
				Week:       week,
				PrM:        prM,
				CpCyc:      cpCyc,
				DoHz:       float64(doHz),
				LockTimeMs: int(binary.LittleEndian.Uint16(m[24:])),
				Cn0:        int(m[26]),
				CpValid:    cpValid,
				CycleSlip:  trkStat&0x08 != 0,
				ArcBreak:   arcBreak,
			},
		})
		emitted++
	}
	return emitted, true
}

func finiteFloat(v float64) bool { return !math.IsNaN(v) && !math.IsInf(v, 0) }

// parseSFRBX converts a UBX-RXM-SFRBX payload to a RawFrame. Layout (F9/M9
// generation): gnssId, svId, sigId, freqId, numWords, reserved, version,
// reserved, then numWords little-endian 32-bit dwrds each holding one native nav
// word right-aligned. Returns nil on a malformed/short payload.
func parseSFRBX(p []byte, source string, recv time.Time) *RawFrame {
	if len(p) < 8 {
		return nil
	}
	gnssID := gnss.GNSSID(p[0])
	svID := int(p[1])
	sigID := int(p[2])
	freqID := int(p[3])
	numWords := int(p[4])
	if numWords == 0 || 8+numWords*4 > len(p) {
		return nil
	}
	words := make([]uint32, numWords)
	for i := 0; i < numWords; i++ {
		words[i] = binary.LittleEndian.Uint32(p[8+i*4:])
	}
	return &RawFrame{
		Recv:   recv,
		Source: source,
		GnssID: gnssID,
		SvID:   svID,
		SigID:  sigID,
		FreqID: freqID,
		Words:  words,
	}
}

// parseMONRF converts a UBX-MON-RF payload (F9+ RF-front-end telemetry) into a
// station-scoped RF sample (docs/DEFENSE-PNT.md §1). Layout: version U1, nBlocks U1,
// reserved U1[2], then nBlocks × 24-byte blocks — blockId U1, flags X1 (bits 0-1 =
// jammingState), antStatus U1, antPower U1, postStatus U4, reserved U1[4], noisePerMS U2,
// agcCnt U2, jamInd U1, … Every length is bounds-checked. Returns nil on a short frame.
func parseMONRF(p []byte, source string, recv time.Time) *RawFrame {
	if len(p) < 4 {
		return nil
	}
	nBlocks := int(p[1])
	if nBlocks == 0 || 4+nBlocks*24 > len(p) {
		return nil
	}
	rf := &RawRF{Bands: make([]RFBand, 0, nBlocks)}
	for i := 0; i < nBlocks; i++ {
		b := p[4+i*24:]
		rf.Bands = append(rf.Bands, RFBand{
			Block:     int(b[0]),
			JamState:  int(b[1] & 0x03),
			AntStatus: int(b[2]),
			// postStatus U4 sits at +4, reserved U1[4] at +8 (12 bytes total,
			// matching the doc comment above), then noisePerMS U2 @+12, agcCnt U2
			// @+14, jamInd U1 @+16 — the previous +14/+16/+20 offsets read agcCnt as
			// noise, jamInd|ofsI<<8 as AGC, and magQ as CWSuppress.
			NoiseLevel: int(binary.LittleEndian.Uint16(b[12:])),
			AGC:        int(binary.LittleEndian.Uint16(b[14:])),
			CWSuppress: int(b[16]), // jamInd (CW-jamming indicator, 0..255)
		})
	}
	return &RawFrame{Source: source, Recv: recv, RF: rf}
}

// parseMONHW converts a legacy UBX-MON-HW payload (60 bytes) into a single-band RF
// sample: noisePerMS U2 @16, agcCnt U2 @18, aStatus U1 @20, flags X1 @22 (jammingState in
// bits 2-3), jamInd U1 @45 (docs/DEFENSE-PNT.md §1). Returns nil on a short frame.
func parseMONHW(p []byte, source string, recv time.Time) *RawFrame {
	if len(p) < 60 {
		return nil
	}
	band := RFBand{
		Block:      0,
		NoiseLevel: int(binary.LittleEndian.Uint16(p[16:])),
		AGC:        int(binary.LittleEndian.Uint16(p[18:])),
		AntStatus:  int(p[20]),
		JamState:   int((p[22] >> 2) & 0x03),
		CWSuppress: int(p[45]),
	}
	return &RawFrame{Source: source, Recv: recv, RF: &RawRF{Bands: []RFBand{band}}}
}

// parseNAVSAT converts a UBX-NAV-SAT payload into per-SV C/N₀ + elevation for the
// C/N₀-vs-elevation spoofing gate (docs/DEFENSE-PNT.md §3). Layout: iTOW U4, version U1,
// numSvs U1, reserved U1[2], then numSvs × 12-byte blocks — gnssId U1, svId U1, cno U1
// (dB-Hz), elev I1 (deg), azim I2, prRes I2, flags X4 (bit 3 = svUsed). Bounds-checked.
func parseNAVSAT(p []byte, source string, recv time.Time) *RawFrame {
	if len(p) < 8 {
		return nil
	}
	numSvs := int(p[5])
	if numSvs == 0 || 8+numSvs*12 > len(p) {
		return nil
	}
	rf := &RawRF{Sats: make([]SatCN0, 0, numSvs)}
	for i := 0; i < numSvs; i++ {
		s := p[8+i*12:]
		flags := binary.LittleEndian.Uint32(s[8:])
		rf.Sats = append(rf.Sats, SatCN0{
			GnssID:  int(s[0]),
			SvID:    int(s[1]),
			Cn0:     int(s[2]),
			ElevDeg: int(int8(s[3])),
			Used:    flags&0x08 != 0,
		})
	}
	return &RawFrame{Source: source, Recv: recv, RF: rf}
}

// syncTo advances the reader until the two-byte sync pattern s1,s2 is consumed.
func syncTo(br *bufio.Reader, s1, s2 byte) error {
	for {
		b, err := br.ReadByte()
		if err != nil {
			return err
		}
		if b != s1 {
			continue
		}
		b2, err := br.ReadByte()
		if err != nil {
			return err
		}
		if b2 == s2 {
			return nil
		}
		// Not the pattern; b2 might itself be s1 — push it back to re-examine.
		if b2 == s1 {
			_ = br.UnreadByte()
		}
	}
}

// fletcher8 computes the UBX 8-bit Fletcher checksum over class, id, the two
// length bytes, and the payload.
func fletcher8(cls, id, lenLo, lenHi byte, body []byte) (ckA, ckB byte) {
	add := func(x byte) {
		ckA += x
		ckB += ckA
	}
	add(cls)
	add(id)
	add(lenLo)
	add(lenHi)
	for _, x := range body {
		add(x)
	}
	return ckA, ckB
}

File: esp32/components/ubx/src/ubx.c
// ubx — UBX framer + GNF1 record emitters. See include/ubx.h.
// The record-body layouts are byte-for-byte matched to ../../../feeder/navfeeder.c
// (emit_sfrbx/monrf/monhw/navsat) and the collector's scanUBX. For supported signals the
// complete records match; gnf1_frame_type documents the ESP32's broader fallback labels
// for unknown signal IDs.

#include "ubx.h"

#include <string.h>

// UBX protocol constants (u-blox interface description).
#define UBX_SYNC1     0xB5
#define UBX_SYNC2     0x62
#define UBX_CLASS_NAV 0x01
#define UBX_CLASS_RXM 0x02
#define UBX_CLASS_MON 0x0A
#define UBX_ID_SFRBX  0x13
#define UBX_ID_NAVSAT 0x35
#define UBX_ID_MONHW  0x09
#define UBX_ID_MONRF  0x38

#define MAX_TELEM_SATS 200 // matches navfeeder MAX_TELEM_SATS and Go maxTelemSats

// framer states
enum { S_SYNC1 = 0, S_SYNC2, S_CLASS, S_ID, S_LEN1, S_LEN2, S_PAYLOAD, S_CK_A, S_CK_B };

// UBX telemetry fields are little-endian on the wire.
static inline uint16_t rd_le16(const uint8_t *b) { return (uint16_t)b[0] | ((uint16_t)b[1] << 8); }
static inline uint32_t rd_le32(const uint8_t *b) {
    return (uint32_t)b[0] | ((uint32_t)b[1] << 8) | ((uint32_t)b[2] << 16) | ((uint32_t)b[3] << 24);
}

void ubx_parser_init(ubx_parser_t *p, ubx_emit_fn emit, ubx_now_fn now, void *ctx)
{
    memset(p, 0, sizeof *p);
    p->emit = emit;
    p->now = now;
    p->ctx = ctx;
    p->state = S_SYNC1;
}

// --- record emitters (payload is the validated UBX message body) -------------------------

// emit_sfrbx: UBX-RXM-SFRBX -> raw-nav record. Layout (F9/M9): gnssId, svId, sigId, freqId,
// numWords, reserved, version, reserved, then numWords little-endian dwrds each holding one
// native nav word. Re-serialize each word big-endian (the collector reads them back BE).
static void emit_sfrbx(ubx_parser_t *p, const uint8_t *payload, uint16_t len)
{
    if (len < 8) return;
    unsigned gnss_id = payload[0], sv_id = payload[1], sig_id = payload[2],
             freq_id = payload[3], num_words = payload[4];
    if (num_words == 0 || 8u + num_words * 4u > len) return;
    unsigned raw_len = num_words * 4u;
    if (raw_len > GNF1_MAX_RAW) return;

    uint8_t *rec = p->scratch;
    for (unsigned i = 0; i < num_words; i++)
        gnf1_be32(rec + GNF1_RECORD_HDR + i * 4, rd_le32(payload + 8 + i * 4));
    gnf1_be64(rec, p->now());
    rec[8]  = (uint8_t)gnss_id;
    rec[9]  = (uint8_t)sv_id;
    rec[10] = (uint8_t)sig_id;
    rec[11] = (uint8_t)freq_id;
    rec[12] = gnf1_frame_type(gnss_id, sig_id);
    p->emit(rec, GNF1_RECORD_HDR + raw_len, p->ctx);
    atomic_fetch_add_explicit(&p->frames_nav, 1, memory_order_relaxed); /* regression fix */
}

// emit_monrf: UBX-MON-RF (F9+) -> JammingStats (0x05). Layout: version U1, nBlocks U1,
// reserved U1[2], then nBlocks x 24-byte blocks (blockId U1, flags X1 [bits0-1=jamState],
// antStatus U1, antPower U1, postStatus U4 @4, reserved U1[4] @8, noisePerMS U2 @12,
// agcCnt U2 @14, jamInd U1 @16 (previously read @14/@16/@20, off by the 2-byte
// antStatus/antPower pair).
static void emit_monrf(ubx_parser_t *p, const uint8_t *payload, uint16_t len)
{
    if (len < 4) return;
    unsigned n_blocks = payload[1];
    if (n_blocks == 0 || 4u + n_blocks * 24u > len) return;
    unsigned body_len = 2u + n_blocks * 8u;
    if (body_len > GNF1_MAX_RAW) return;

    uint8_t *rec = p->scratch;
    uint8_t *body = rec + GNF1_RECORD_HDR;
    body[0] = GNF1_TELEM_VERSION;
    body[1] = (uint8_t)n_blocks;
    for (unsigned i = 0; i < n_blocks; i++) {
        const uint8_t *b = payload + 4 + i * 24;
        unsigned o = 2 + i * 8;
        body[o]     = b[0];                          // blockId
        gnf1_be16(body + o + 1, rd_le16(b + 14));    // agcCnt
        gnf1_be16(body + o + 3, rd_le16(b + 12));    // noisePerMS
        body[o + 5] = b[16];                         // jamInd (CW)
        body[o + 6] = (uint8_t)(b[1] & 0x03);        // jammingState
        body[o + 7] = b[2];                          // antStatus
    }
    gnf1_be64(rec, p->now());
    rec[8] = rec[9] = rec[10] = rec[11] = 0;
    rec[12] = GNF1_T_JAMMING;
    p->emit(rec, GNF1_RECORD_HDR + body_len, p->ctx);
    atomic_fetch_add_explicit(&p->frames_telem, 1, memory_order_relaxed); /* regression fix */
}

// emit_monhw: legacy UBX-MON-HW (60 bytes) -> single-band JammingStats. noisePerMS U2 @16,
// agcCnt U2 @18, aStatus U1 @20, flags X1 @22 (jammingState bits 2-3), jamInd U1 @45.
static void emit_monhw(ubx_parser_t *p, const uint8_t *payload, uint16_t len)
{
    if (len < 60) return;
    uint8_t *rec = p->scratch;
    uint8_t *body = rec + GNF1_RECORD_HDR;
    body[0] = GNF1_TELEM_VERSION;
    body[1] = 1;
    body[2] = 0;                                     // block 0
    gnf1_be16(body + 3, rd_le16(payload + 18));      // agcCnt
    gnf1_be16(body + 5, rd_le16(payload + 16));      // noisePerMS
    body[7] = payload[45];                           // jamInd (CW)
    body[8] = (uint8_t)((payload[22] >> 2) & 0x03);  // jammingState
    body[9] = payload[20];                           // aStatus
    gnf1_be64(rec, p->now());
    rec[8] = rec[9] = rec[10] = rec[11] = 0;
    rec[12] = GNF1_T_JAMMING;
    p->emit(rec, GNF1_RECORD_HDR + 10, p->ctx);
    atomic_fetch_add_explicit(&p->frames_telem, 1, memory_order_relaxed); /* regression fix */
}

// emit_navsat: UBX-NAV-SAT -> ReceptionData (0x01) for the C/N0-vs-elevation spoof gate.
// Layout: iTOW U4, version U1, numSvs U1 @5, reserved U1[2], then numSvs x 12-byte blocks
// (gnssId U1, svId U1, cno U1 @2, elev I1 @3, ..., flags X4 @8 [bit3=svUsed]).
static void emit_navsat(ubx_parser_t *p, const uint8_t *payload, uint16_t len)
{
    if (len < 8) return;
    unsigned num_svs = payload[5];
    if (num_svs == 0 || 8u + num_svs * 12u > len) return;
    unsigned n = num_svs > MAX_TELEM_SATS ? MAX_TELEM_SATS : num_svs;

    uint8_t *rec = p->scratch;
    uint8_t *body = rec + GNF1_RECORD_HDR;
    body[0] = GNF1_TELEM_VERSION;
    gnf1_be16(body + 1, (uint16_t)n);
    for (unsigned i = 0; i < n; i++) {
        const uint8_t *s = payload + 8 + i * 12;
        unsigned o = 3 + i * 5;
        body[o]     = s[0];                          // gnssId
        body[o + 1] = s[1];                          // svId
        body[o + 2] = s[2];                          // cno
        body[o + 3] = s[3];                          // elev (I1, verbatim)
        body[o + 4] = (rd_le32(s + 8) & 0x08) ? 0x01 : 0x00; // svUsed
    }
    gnf1_be64(rec, p->now());
    rec[8] = rec[9] = rec[10] = rec[11] = 0;
    rec[12] = GNF1_T_RECEPTION;
    p->emit(rec, GNF1_RECORD_HDR + 3u + n * 5u, p->ctx);
    atomic_fetch_add_explicit(&p->frames_telem, 1, memory_order_relaxed); /* regression fix */
}

// dispatch routes a checksum-valid message to its emitter.
static void dispatch(ubx_parser_t *p)
{
    if (p->cls == UBX_CLASS_RXM && p->id == UBX_ID_SFRBX)      emit_sfrbx(p, p->payload, p->len);
    else if (p->cls == UBX_CLASS_MON && p->id == UBX_ID_MONRF) emit_monrf(p, p->payload, p->len);
    else if (p->cls == UBX_CLASS_MON && p->id == UBX_ID_MONHW) emit_monhw(p, p->payload, p->len);
    else if (p->cls == UBX_CLASS_NAV && p->id == UBX_ID_NAVSAT) emit_navsat(p, p->payload, p->len);
}

// ck accumulates one byte into the running Fletcher-8 checksum.
static inline void ck(ubx_parser_t *p, uint8_t b) { p->ck_a += b; p->ck_b += p->ck_a; }

void ubx_parser_feed(ubx_parser_t *p, const uint8_t *data, size_t len)
{
    for (size_t i = 0; i < len; i++) {
        uint8_t b = data[i];
        switch (p->state) {
        case S_SYNC1:
            if (b == UBX_SYNC1) p->state = S_SYNC2;
            break;
        case S_SYNC2:
            // A lone 0xB5 followed by another 0xB5 keeps the second as a fresh candidate.
            if (b == UBX_SYNC2) p->state = S_CLASS;
            else if (b != UBX_SYNC1) p->state = S_SYNC1;
            break;
        case S_CLASS:
            p->cls = b; p->ck_a = p->ck_b = 0; ck(p, b); p->state = S_ID;
            break;
        case S_ID:
            p->id = b; ck(p, b); p->state = S_LEN1;
            break;
        case S_LEN1:
            p->len = b; ck(p, b); p->state = S_LEN2;
            break;
        case S_LEN2:
            p->len |= (uint16_t)b << 8; ck(p, b);
            if (p->len > UBX_MAX_PAYLOAD) { atomic_fetch_add_explicit(&p->oversize, 1, memory_order_relaxed); p->state = S_SYNC1; break; }
            p->idx = 0;
            p->state = p->len ? S_PAYLOAD : S_CK_A;
            break;
        case S_PAYLOAD:
            p->payload[p->idx++] = b; ck(p, b);
            if (p->idx >= p->len) p->state = S_CK_A;
            break;
        case S_CK_A:
            p->exp_a = b; p->state = S_CK_B;
            break;
        case S_CK_B:
            p->exp_b = b;
            if (p->ck_a == p->exp_a && p->ck_b == p->exp_b) dispatch(p);
            else atomic_fetch_add_explicit(&p->bad_checksum, 1, memory_order_relaxed); /* regression fix */
            p->state = S_SYNC1;
            break;
        default:
            p->state = S_SYNC1;
            break;
        }
    }
}
